static/app/utils/profiling/profile/chromeTraceProfile.tsx

```diff
@@ -0,0 +1,286 @@
+
+  // We dont care about other events besides begin, end, instant events and metadata events
+  const timelineEvents = events.filter(
+    e => e.ph === 'B' || e.ph === 'E' || e.ph === 'X' || e.ph === 'M'
+  );
+
+  // @TODO use a heap so we dont need to sort this again afterwards
+  const beginQueue: Array<ChromeTrace.Event> = [];
+  const endQueue: Array<ChromeTrace.Event> = [];
+
+  for (let i = 0; i < timelineEvents.length; i++) {
+    const event = timelineEvents[i];
+
+    // M events are not pushed to the queue, we just store their information
+    if (event.ph === 'M') {
+      if (event.name === 'thread_name' && typeof event.args.name === 'string') {
+        threadName = `${event.args.name} (${threadId})`;
+        continue;
+      }
+
+      if (event.name === 'process_name' && typeof event.args.name === 'string') {
+        processName = `${event.args.name} (${processId})`;
+        continue;
+      }
+    }
+
+    // B, E and X events are pushed to the timeline. We transform all X events into
+    // B and E event, so that they can be pushed onto the queue and handled
+    if (event.ph === 'B') {
+      beginQueue.push(event);
+      continue;
+    }
+
+    if (event.ph === 'E') {
+      endQueue.push(event);
+      continue;
+    }
+
+    if (event.ph === 'X') {
+      if (typeof event.dur === 'number' || typeof event.tdur === 'number') {
+        beginQueue.push({...event, ph: 'B'});
+        endQueue.push({...event, ph: 'E', ts: event.ts + (event.dur ?? event.tdur ?? 0)});
+        continue;
+      }
+    }
+  }
+
+  beginQueue.sort(reverseChronologicalSort);
+  endQueue.sort(reverseChronologicalSort);
+
+  if (!beginQueue.length) {
+    throw new Error('Profile does not contain any frame events');
+  }
+
+  const firstTimestamp = beginQueue[beginQueue.length - 1].ts;
+
+  if (typeof firstTimestamp !== 'number') {
+    throw new Error('First begin event contains no timestamp');
+  }
+
+  const profile = new ChromeTraceProfile(
+    0,
+    0,
+    0,
+    `${processName}: ${threadName}`,
+    'milliseconds'
+  );
+
+  const stack: ChromeTrace.Event[] = [];
+  const frameCache = new Map<string, Frame>();
+
+  while (beginQueue.length > 0 || endQueue.length > 0) {
+    const next = getNextQueue(beginQueue, endQueue);
+
+    if (next === 'B') {
+      const item = beginQueue.pop();
+      if (!item) {
+        throw new Error('Nothing to take from begin queue');
+      }
+
+      const frameInfo = createFrameInfoFromEvent(item);
+
+      if (!frameCache.has(frameInfo.key)) {
+        frameCache.set(frameInfo.key, new Frame(frameInfo));
+      }
+
+      const frame = frameCache.get(frameInfo.key)!;
+      profile.enterFrame(frame, item.ts - firstTimestamp);
+      stack.push(item);
+      continue;
+    }
+
+    if (next === 'E') {
+      const item = endQueue.pop()!;
+      let frameInfo = createFrameInfoFromEvent(item);
+      const topFrameInfo = createFrameInfoFromEvent(stack[stack.length - 1]);
```

Comment:
`stack[stack.length - 1]` can be `undefined` in a bad/corrupt profile right? Do we need to guard against that?

Reply:
Nice catch, I will add a check and throw if that's the case